packages/php-wasm/universal/src/lib/php.ts

```diff
@@ -755,80 +852,157 @@ export class PHP implements Disposable {
-		 * get crashes and unhandled promise rejections without any useful error
-		 * messages or stack traces.
-		 */
 		let errorListener: any;
-		try {
-			// eslint-disable-next-line no-async-promise-executor
-			exitCode = await new Promise<number>((resolve, reject) => {
-				errorListener = (e: ErrorEvent) => {
-					logger.error(e);
-					logger.error(e.error);
-					const rethrown = new Error('Rethrown');
-					rethrown.cause = e.error;
-					(rethrown as any).betterMessage = e.message;
-					reject(rethrown);
-				};
-				this.#wasmErrorsTarget?.addEventListener(
+
+		const runExecutionFunction = async () => {
+			try {
+				/*
+				 * Emscripten throws WASM failures outside of the promise chain so we need
+				 * to listen for them here and rethrow in the correct context. Otherwise we
+				 * get crashes and unhandled promise rejections without any useful error
+				 * messages or meaningful stack traces.
+				 */
+				return await Promise.race([
+					executionFn(),
+					new Promise((_, reject) => {
+						errorListener = (e: ErrorEvent) => {
+							// @TODO: Why are we logging things so eagerly? Shouldn't
+							//        the caller handle that? There's so much noise in
+							//        the console.
+							// logger.error(e);
+							// logger.error(e.error);
+							if (!isExitCode(e.error)) {
+								const rethrown = new Error('Rethrown');
+								rethrown.cause = e.error;
+								(rethrown as any).betterMessage = e.message;
+								reject(rethrown);
+							}
+						};
+						this.#wasmErrorsTarget?.addEventListener(
+							'error',
+							errorListener,
+							{ once: true }
+						);
+					}),
+				]);
+			} catch (e) {
+				/**
+				 * Emscripten sometimes communicates program exit as an error. Let's
+				 * turn exit code errors into integers again.
+				 */
+				if (isExitCode(e)) {
+					return e.exitCode;
+				}
+
+				stdout.controller.error(e);
+				stderr.controller.error(e);
+				headers.controller.error(e);
+				streamsClosed = true;
+
+				/**
+				 * A non-exit-code error means an irrecoverable crash. Let's make
+				 * it very clear to the consumers of this API – every method
+				 * call on this PHP instance will throw an error from now on.
+				 */
+				for (const name in this) {
+					if (typeof this[name] === 'function') {
+						(this as any)[name] = () => {
+							throw new Error(
+								`PHP runtime has crashed – see the earlier error for details.`
+							);
+						};
+					}
+				}
+				(this as any).functionsMaybeMissingFromAsyncify =
+					getFunctionsMaybeMissingFromAsyncify();
+
+				const err = e as Error;
+				const message = (
+					'betterMessage' in err ? err.betterMessage : err.message
+				) as string;
+
+				const rethrown = new Error(message);
+				rethrown.cause = err;
+				// @TODO: Why are we logging things so eagerly? Shouldn't
+				//        the caller handle that? There's so much noise in
+				//        the console.
+				// logger.error(rethrown);
```

Comment:
I found the [original PR](https://github.com/WordPress/wordpress-playground/pull/1315/files#diff-b33e1e6834fb94ac2cb5338e3ae17c60abf7ac88cb057189a1915531e55f7243) and it looks like I just replaced all `console.error` calls with `logger.error`, so this might be a good time to rethink if we really need these log entries.